chore(transformer): add legacy decorator tests that come from `TypeScript`

Dunqing opened the pull request:
Part of #8614 

All tests copy over from https://github.com/microsoft/TypeScript/blob/8da951cbb629b648753454872df4e1754982aef1/tests/cases/conformance/decorators/class, in addition, several multi-file tests were also adjusted.

The output is generated by
```ts
transpileModule(
  source,
  {
    compilerOptions: {
      target: 'esnext',
      experimentalDecorators: true,
      noEmitHelpers: true
    },
  }
)
```
and replace all `__decorate` with `babelHelpers.decorate` and `__param` with `babelHelpers.decorateParam`

Dunqing commented:
* **#8692** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8692?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8692?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8692).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
When ever you decide to merge.

overlookmotel commented on tasks/transform_conformance/tests/legacy-decorators/test/fixtures/index.js:
```diff
@@ -0,0 +1,41 @@
+// }
+async function main() {
+  const files = await readdir(__dirname, {recursive: true});
+  files.map((file) => join(__dirname, file) ).forEach(async file => {
```
This is very Rust-y Javascript!

Personally, I'd just do it with a `for` loop (run in series) because it'd be simpler to understand, and reading/writing 70 files simultaneously probably isn't very efficient anyway - it probably overloads the file system.

But if you want to keep it async, you should `await` the resulting promises, or they're just "firing off into space", and the function returns before they're all done. I'm not sure if errors would get handled.

```js
await Promise.all(
  files.map((file) => join(__dirname, file) )
    // NOTE: `.map` not `.forEach`
    .map(async file => {
      // Do async stuff
    })
);
```

I guess it doesn't matter much since all the results are checked into git, but since it's committed to the repo...

Dunqing replied:
Oh, I mistakenly committed it, I initially thought that would never use it again, so writing it casually. But you said right, it would be better if we can retrace in the future. I will re-implement it.

overlookmotel commented on tasks/transform_conformance/tests/legacy-decorators/test/fixtures/index.js:
```diff
@@ -0,0 +1,41 @@
+import {  } from 'fs';
```
I think this line is extraneous.

```suggestion
```

overlookmotel reviewed (commented):
Aside from my comments about the JS, there's only one thing... could you please add a comment to the top of the fixtures updater, with the URL where the original fixtures come from? And also note what kind of modifications you made to the fixtures, aside from what this script does?

If we find bugs many months down the line, it'd be helpful to have this info if we need to retrace our steps...

overlookmotel commented:
> Aside from my comments about the JS, there's only one thing... could you please add a comment to the top of the fixtures updater, with the URL where the original fixtures come from? And also note what kind of modifications you made to the fixtures, aside from what this script does?

Ignore my last comment! I just saw that info's all in the comment on the transformer added in #8614.

overlookmotel reviewed (approved):
Great!

Please go ahead and merge once conformance task fail is fixed.

Dunqing commented:
### Merge activity

* **Feb 7, 9:47 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
chore(transformer): add legacy decorator tests that come from `TypeScript` (#8692)

Part of #8614

All tests copy over from https://github.com/microsoft/TypeScript/blob/8da951cbb629b648753454872df4e1754982aef1/tests/cases/conformance/decorators/class, in addition, several multi-file tests were also adjusted.

The output is generated by
```ts
transpileModule(
  source,
  {
    compilerOptions: {
      target: 'esnext',
      experimentalDecorators: true,
      noEmitHelpers: true
    },
  }
)
```
and replace all `__decorate` with `babelHelpers.decorate` and `__param` with `babelHelpers.decorateParam`
